OPENNLP-1267 -- add a ProbingLanguageDetector that can stop early.

tballison opened the pull request:
Thank you for contributing to Apache OpenNLP.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ *] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ *] Does your PR title start with OPENNLP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ *] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ *] Is your initial contribution a single, squashed commit?

### For code changes:
- [ *] Have you ensured that the full suite of tests is executed via mvn clean install at the root opennlp folder? 
- [ *] Have you written or updated unit tests to verify your changes?
- [ *] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ *] If applicable, have you updated the LICENSE file, including the main LICENSE file in opennlp folder?
- [ *] If applicable, have you updated the NOTICE file, including the main NOTICE file found in opennlp folder?

### For documentation related changes:
- [ *] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

tballison commented:
I'll squash the second commit on Monday.  Let me know what you think.

tballison commented:
Correction...I'll probably make some further refinements on Monday and then squash.  No need to review now.

tballison commented:
Updates made and squashed.  Let me know what you think.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java:
```diff
@@ -0,0 +1,412 @@
+ * </p>
+ *
+ */
+public class ProbingLanguageDetectorME extends LanguageDetectorME {
```
The probing should be added to the LanguageDetectorME directly. There could be two style of methods, the one which it has currently, and a second method which does the probing like you implemented it. The second method could also have a different return type in case that makes it easier to communicate the detected languages to the caller (e.g. return additional information).

tballison replied:
K.  Sounds good.

>  (e.g. return additional information)
e.g. how much of the text was probed?

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java:
```diff
@@ -0,0 +1,412 @@
+   */
+  private int maxLength = DEFAULT_MAX_LENGTH;
+
+  private CharSequenceNormalizer normalizer;
```
The normalizer is already part of the LanguageDetectorME and shouldn't be created here, also it shouldn't be possible to modify it. The normalizer is declared at training time, and also can only be changed at training time. The reason for that is that the normalizer has an influence on the features and these have to match the ones at training time.

tballison replied:
Y.  Got it.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java:
```diff
@@ -0,0 +1,412 @@
+  public Language[] predictLanguages(CharSequence content) {
+    //list of the languages that received the highest
+    //confidence over the last n chunk detections
+    LinkedList<Language[]> predictions = new LinkedList();
```
Should be List<Language[]> instead.

tballison replied:
y.  Thank you.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java:
```diff
@@ -0,0 +1,412 @@
+    int start = 0;//where to start the next chunk in codepoints
+    Language[] currPredictions = null;
+    //cache ngram counts across chunks
+    Map<String, MutableInt> ngramCounts = new HashMap<>();
```
Can the existing predict method be called here? It should not be duplicated here, if there is something wrong with the way the ngrams are generated the place where this originally happens should be fixed. The generation of the features is done in the Context Generator, and could also be swapped if there are different needs, even by the user at training time.

tballison replied:
Sounds good.  What I was trying to get away from was the requirement that the ContextGenerator process the full chunk and return String[].  It would be better...at some point... to have the ContextGenerator implement Iterator<String> to enable processing as little as possible.

I think in favor of this incremental, non-breaking change, I can leave the ContextGenerator as is and use the existing predict method.

kottmann replied:
Also not using that ContextGenerator is an option you can implement your own, and then use it instead for training and inference. I see that th einterface is an issue, but this goes down to the model, and other approaches like using hash features would probably the better way to speed it up.

tballison replied:
Y, agreed on hashing.  

From a probing perspective, moving to `Iterator<String>` would be better than getting all the ngrams from a chunk returned as String[].  One option would be to subclass ContextGenerator with an interface that supports `Iterator<String>` or we could simply add a check in the prober to see if the ContextGenerator happens to support `Iterator<String>`.

kottmann commented on opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java:
```diff
@@ -0,0 +1,412 @@
+   * Throws {@link IllegalArgumentException} if &lt; 0.0
+   * @param minDiff
+   */
+  public void setMinDiff(double minDiff) {
```
It would be better if these are passed directly to the predict method. To allow certain features, such as multi threading, and serializing objects it is better if the component itself is immutable.

tballison replied:
+1.  Are you ok if I create a config object that can optionally be passed in to `detect`?

kottmann replied:
I think there should be a simple version that kind of works out of the box.

Anyway, we could overload the method with one which takes more parameters, but if there are too many parameters it would be better to pass a config object to it. Or see if we somehow can eliminate some of the parameters.

tballison replied:
How's it look now? with `probingPredictLanguages(CharSequence txt)`

tballison commented:
When I make the updates in response to the discussion above, should I squash/force push or do you want to see the history from the initial PR?  Thank you, again!

tballison commented:
I refactored the probing parts into LanguageDetectorME.  I removed the optimization that would allow for early stopping within a chunk for simplicity and to rely on the current {{ContextGenerator}} api.

I haven't squashed this because I do want to be able to revert if I misunderstood any of the above guidance.  I can squash before the final commit.

Let me know what you think.

kottmann commented:
We merge only one commit. So things would need to be squashed either by you or by GitHub when we merge. I personally prefer it when you could just send us one commit, but it doesn't matter much anyway because the GH button can do that for us.

tballison commented:
bq. We merge only one commit.

Makes sense.  I'll squash when you're ok with the PR.
